README.md
__FEATURED TODO__
+ Development testing
+ Production testing

__LAST CHANGES__ (other changes see `docs/changelog.txt`)

> __20.01.2019__
+ Testing some cases related c/build/click
+ Documentation for modules and thin places
+ Removin database descriptions when update flows
+ Preparing to release
+ Reformated some parts
+ Single click implementation (viewer)
+ Release 2.0.5alpha

__INSTALLATION__

1. clone this repository like this:
> `$ git clone https://github.com/CpaMonstr/metatds.git`

2. in folder `metacds` run following comand
> `$ dep init`

if you have no `dep` installed, just use
> `$ curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh`

3. edit configuration file `settings.ini` to make sure
your settings are correct

4. Run the following command in project folder

for Linux
> `$ ./make.sh` 

or for Windows
> `$ make.cmd`

it will build project and automaticly run it.

Additional information you can find in /docs folder,
there are format specifications and usage examples


docs/changelog.md
__MetaTDS v.2 CHANGE LOG__

> __04.12.2018__
+ Core functions and logic
+ GIN-GO Framework in use
+ Redirect functions (old format)

> __05.12.2018__
+ Config `settings.ini` file reader / writer
+ Partial ports from LUA language TDS v.1
+ Redirect functions (new format)

> __06.12.2018__
+ Self format for RedisDB
+ Self format for respond
+ Debug utilities

> __07.12.2018__
+ Flow add/update API via HTTP
+ Click info API + formats for back-requests from prelands
+ Click generation
+ Types and methods for models

> __09.12.2018__
+ GIN-GO Framework changed to ECHO Minimalist
+ Functional dropped to file modules
+ Optimizing and refactoring some parts
+ test build for Ubuntu 18.04

> __10.12.2018__
+ Support for backcalls from Prelands
+ Extended info by all Clicks

> __11.12.2018__
+ JSON format support for all features

> __12.12.2018__
+ JSON format requests / responds
+ DEP package manager support
+ test build via `gccgo` compiler for Ubuntu 18.04

> __14.12.2018__
+ Github.com working repo + branches
+ Additional features for HIT export
+ Functional dropped to packages

> __16.12.2018__
+ Telegram notifications for statistic / usage
+ Tested with GCCGO compiler over Ubuntu 18.04
+ Refactored directory structure
+ Minor fixes

> __17.12.2018__
+ Time bench and statistics
+ Dynamic reload .ini file to get settings online
+ Sending custom stats to Telegram Bot
+ Import flows into Redis

> __18.12.2018__
+ Failover refusing connections withing 5sec. IN/OUT
+ Redis crash and recover support
+ Import Flows to Redis
+ Save data to tests-log file `./tdstest.log`
+ Advanced setting cookie
+ Time usage statistics

> __19.12.2018__
+ fx 30 seconds drop connection
+ Recovering when Cookie reading
+ Redis temp func removeing objects 1h/rate
+ Export list of clicks 
+ Creating click with path to append
  ClickHouse export format
  
> __23.12.2018__
+ Improvements for Telegram Bot
+ Saving clicks from RedisDB to a files dir ./clicks/XXXXXXXXXX.json
+ Improvements for Statistics and memory usage params
+ Garbage collector implements
+ Mass testing analasys and fixes bugs depends on

> __24.12.2018__
+ for more stabiliy and graceful deployment
  `settings.ini` is not in a package anymore, now after deployment
  you should rename `settings.dev` or `settings.prod` to `settings.ini
+ fixed bug with nil pointer when telegram can't connect to proxy
+ added library for memory/system monitoring  

> __25.12.2018__
+ testing issues (too many open files)
+ count open files within statistic
+ absolute file path for `settings.ini`
+ added service configuration for system.d (thanks to `@pfirsov`) 

> __26.12.2018__
+ Sending to API
+ Re-sending in case of failure to API
+ Log files optional to STDOUT 
+ Improvements for `system.d` start script
+ Statistics improvements + average time for respond

> __27.12.2018__
+ Recieving flows from API for startup and for work
+ Re-sending in case of failure to API (clicks) tested
+ Additional time stamp saved to `last.update.time` 
+ Improvements for `system.d` start script
+ Statistics improvements + average time for respond
+ Options for flows and clicks @`settings.ini`

> __30.12.2018__
+ Improved performance for responsing to context
+ Additional statistics /stat, /conf
+ Small refactoring
+ Additional configurable settings @`settings.ini`
+ Production testing

> __31.12.2018 22:29 (коммит новогодний)__
+ Improved average time calculatin
+ Improved saving click with redis HMSET
+ Improved getting infos from all triiggers
  lands / prelands / flows / clicks
+ reused profiling information
+ Humanize
+ fx with count of flows

> __01.01.2019__
+ tests to increase performance of randomization
+ additional configuration & management tools
+ memory profiling
+ cpu profiling
+ string generator fix
+ resend statistic send and file saving refactoring

> __02.01.2019__
+ Multi deleting from Redis by exec One command
+ File I/O reading improvements@go routines

> __03.01.2019__
+ Multi deleting from Redis by exec One command
+ File I/O reading improvements@go routines
+ final testing on high-load with concurency 10000rps

> __06.01.2019__
+ Multi deleting from Redis improvements
+ Reconstruced URL helper
+ final testing on high-load with concurency 1000rps/1billion requests

> __07.01.2019__
+ Improved performance for importing flows
+ Additional statistics /stat, /conf
+ Small refactoring for init.go
+ Additional configurable settings @`settings.ini`
+ Production testing on +10000 rps

> __08.01.2019__
+ Improved statistics for average RPS
+ Advanced template based stat output
+ Additional params to watch for stat
+ fx

> __09.01.2019__
+ Additional field counters for clicks and flows
+ Small refactoring for models/click.go...flow.go
+ Production testing on +1000/+5000/+10000 rps

> __11.01.2019__
+ Sending statistic improvements
+ Integration with API metacpa.ru/api
+ Integration tests
+ SSL/TLS keys/certs support

> __13.01.2019__
+ Imporved building a click by parameters
+ Optimization for reading flow infos
+ Reformating some code + fx
+ Preparing to release

> __14.01.2019__
+ Imporved statistics template
+ Optimization statistics
+ Preparing to release

> __15.01.2019__
+ Optimized all flows current/new work with
  new params flow_hash/click_hash
+ Performing to load current clicks to redis for
  direct flows
+ URL reformatting /c /r
+ Preparing to release

> __16.01.2019__
+ Optimized settings to USE SSL yes|no  
+ Settings to backup clicks yes|no 
+ Preparing to release
+ fx settings

> __17.01.2019__
+ FX for landing_id/prelanding_id replace within template

> __18.01.2019__
+ Testing some cases related c/build/click
+ Testing with some issues about X-Real-IP
  User-Agent (non-saved fields)
+ Preparing to release
+ fx settings

tds.go
/****************************************************************************************************
*
* Main TDS module, special for Meta CPA, Ltd.
* by Michael S. Merzlyakov AFKA predator_pc@02122018
* version v2.0.3
*
* created at 04122018
* last edit: 16122018
*
* usage: $ tds run
*
*****************************************************************************************************/

package main

import (
	"bytes"
	"github.com/labstack/echo/middleware"
	"github.com/labstack/gommon/log"
	"github.com/sevenNt/echo-pprof"
	"golang.org/x/crypto/acme/autocert"
	"io"
	"net/http"
	"os"
	"runtime"
	"runtime/debug"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/labstack/echo"
	"metatds/config"
	"metatds/models"
	"metatds/utils"
)

const tdsModuleName = "tds.go"
const timestampFile = "last.update.time"
const defaultStartOfEpoch = "946684800"

// Карта ключей которые мы хотим получить
var keyMap = []string{"flow_hash", "click_hash", "sub1", "sub2", "sub3", "sub4", "sub5", "format", "f",
	"click_id", "flow_id", "prelanding_id", "landing_id"} // support for old version of TDS

// пиксель для тестов
var pixel = []byte(`data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z/C/HgAGgwJ/lK3Q6wAAAABJRU5ErkJggg==`)

// Тип для полного потока с кликами
type InfoData struct {
	Flow  models.FlowData  // модель потока
	Click models.ClickData // модель клика
}

//
// Main GO package handler startup and settings
//
func main() {
	// let's update our flows on start
	UpdateFlowsListChan()

	// Echo instance
	router := echo.New()

	//router.AutoTLSManager.HostPolicy = autocert.HostWhitelist("<DOMAIN>")
	// Cache certificates
	router.AutoTLSManager.Cache = autocert.DirCache("./.cache")

	// Middleware
	//router.Use(middleware.Logger())
	router.Use(middleware.Recover())

	//avoid chrome to request favicon
	router.GET("/favicon.ico", func(c echo.Context) error {
		return c.Blob(200, "image/png", pixel)
		//return c.String(404, "not found") //nothing
	})

	// Routes
	router.GET("/r", flowHandler)
	router.GET("/r/", flowHandler)
	router.GET("/r/:flow_hash", flowHandler)
	router.GET("/r/:flow_hash/", flowHandler)
	router.GET("/r/:flow_hash/:sub1", flowHandler)
	router.GET("/r/:flow_hash/:sub1/", flowHandler)
	router.GET("/r/:flow_hash/:sub1/:sub2", flowHandler)
	router.GET("/r/:flow_hash/:sub1/:sub2/", flowHandler)
	router.GET("/r/:flow_hash/:sub1/:sub2/:sub3", flowHandler)
	router.GET("/r/:flow_hash/:sub1/:sub2/:sub3/", flowHandler)
	router.GET("/r/:flow_hash/:sub1/:sub2/:sub3/:sub4", flowHandler)
	router.GET("/r/:flow_hash/:sub1/:sub2/:sub3/:sub4/", flowHandler)
	router.GET("/r/:flow_hash/:sub1/:sub2/:sub3/:sub4/:sub5", flowHandler)
	router.GET("/r/:flow_hash/:sub1/:sub2/:sub3/:sub4/:sub5/", flowHandler)
	// Routes
	router.GET("/c", clickBuild)
	router.GET("/c/", clickBuild)
	router.GET("/c/:flow_hash", clickBuild)
	router.GET("/c/:flow_hash/", clickBuild)
	router.GET("/c/:flow_hash/:click_hash", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/:sub2", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/:sub2/", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/:sub2/:sub3", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/:sub2/:sub3/", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/:sub2/:sub3/:sub4", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/:sub2/:sub3/:sub4/", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/:sub2/:sub3/:sub4/:sub5", clickBuild)
	router.GET("/c/:flow_hash/:click_hash/:land_id/:preland_id/:sub1/:sub2/:sub3/:sub4/:sub5/", clickBuild)

	// Routes
	router.GET("/", flowHandler)
	router.GET("/:flow_hash", flowHandler)
	router.GET("/:flow_hash/", flowHandler)
	router.GET("/:flow_hash/:sub1", flowHandler)
	router.GET("/:flow_hash/:sub1/", flowHandler)
	router.GET("/:flow_hash/:sub1/:sub2", flowHandler)
	router.GET("/:flow_hash/:sub1/:sub2/", flowHandler)
	router.GET("/:flow_hash/:sub1/:sub2/:sub3", flowHandler)
	router.GET("/:flow_hash/:sub1/:sub2/:sub3/", flowHandler)
	router.GET("/:flow_hash/:sub1/:sub2/:sub3/:sub4", flowHandler)
	router.GET("/:flow_hash/:sub1/:sub2/:sub3/:sub4/", flowHandler)
	router.GET("/:flow_hash/:sub1/:sub2/:sub3/:sub4/:sub5", flowHandler)
	router.GET("/:flow_hash/:sub1/:sub2/:sub3/:sub4/:sub5/", flowHandler)

	router.GET("/info/:click_hash", clickHandler)

	router.GET("/stat", GetSystemStatHandler)
	router.GET("/extstat", GetSystemExtendedStatHandler)
	router.GET("/conf", GetSystemConfHandler)

	router.GET("/free", func(c echo.Context) error {
		debug.FreeOSMemory()
		runtime.GC()
		return c.String(200, "ok")
	})

	router.GET("/memstat", func(c echo.Context) error {

		m := utils.MemMonitor()
		// Just encode to json and print
		return c.String(200, utils.JSONPretty(m))
	})

	customServer := &http.Server{
		Addr:         ":" + strconv.Itoa(config.Cfg.General.Port),
		ReadTimeout:  time.Duration(1+config.Cfg.General.HTTPTimeout) * time.Second,
		WriteTimeout: time.Duration(1+config.Cfg.General.HTTPTimeout) * time.Second,
		IdleTimeout:  time.Duration(1+config.Cfg.General.HTTPTimeout) * time.Second,
	}

	customServer.SetKeepAlivesEnabled(false)
	router.HideBanner = true
	router.Logger.SetLevel(log.OFF)

	if config.Cfg.Debug.Level > 0 {
		echopprof.Wrap(router)
	}

	// router.AutoTLSManager
	if config.Cfg.General.SSL {
		go func() {
			router.Logger.Fatal(router.StartTLS(":443", config.Cfg.General.SSLCert, config.Cfg.General.SSLKey))
		}()
	}

	// run router
	if config.Cfg.General.Port != 0 {
		// regular server
		// router.Logger.Fatal(router.Start(":" + strconv.Itoa(config.Cfg.General.Port)))

		// custom server
		router.Logger.Fatal(router.StartServer(customServer))
	} else {
		//exit if not
		panic("[ERROR] Failed to obtain server port from settings.ini")
	}

}

func GetSystemConfHandler(c echo.Context) error {
	agent := c.Request().UserAgent()
	// защита от долбоебов
	if agent == "MetaDevAgent" {
		text := config.GetSystemConfiguration()
		return c.HTML(200,
			"<html><head><title>TDS System statistics</title><script>"+
				"setInterval(function(){window.location.reload(true)},5000);"+
				"</script></head><body><pre>"+
				text+
				"</pre></body></html>")

	} else {
		return c.String(404, "Not found on server")
	}
}

func GetSystemStatHandler(c echo.Context) error {
	agent := c.Request().UserAgent()
	// защита от долбоебов
	if agent == "MetaDevAgent" {
		text := config.GetSystemStatistics()
		return c.HTML(200,
			"<html><head><title>TDS System statistics</title><script>"+
				"setInterval(function(){window.location.reload(true)},10000);"+
				"</script></head><body><pre>"+
				text+
				"</pre></body></html>")

	} else {
		return c.String(404, "Not found on server")
	}
}

func GetSystemExtendedStatHandler(c echo.Context) error {
	var dataForGraph string
	var dataKeys []int

	// Getting current stats 1 column
	text := config.GetSystemStatistics()

	// Reading file with template for graphic
	w := bytes.NewBuffer(nil)
	file, _ := os.Open("tmpl/sysstat.tmpl")
	_, _ = io.Copy(w, file)
	_ = file.Close()

	// Getting current stats from Redis saved before
	dataFromRedis, _ := config.Redisdb.HGetAll("SystemStatistic").Result()

	for i, _ := range dataFromRedis {
		convertedID, _ := strconv.Atoi(i)
		dataKeys = append(dataKeys, convertedID)
	}

	// needs to be sorted casue of HGetALL output issue
	sort.Ints(dataKeys)

	for _, item := range dataKeys {
		convertedID := strconv.Itoa(item)
		dataForGraph += dataFromRedis[convertedID] + ",\n"
	}

	// Just encode to json and print
	monitor := utils.MemMonitor()
	memoryStats := utils.JSONPretty(monitor)

	// getting template from var
	template := w.String()
	result := strings.Replace(template, "{{MEM}}", memoryStats, -1)
	result = strings.Replace(result, "{{DATA}}", dataForGraph, -1)
	result = strings.Replace(result, "{{SYSSTAT}}", text, -1)

	// returning resulting page
	return c.HTML(200, result)
}
